File: system_control/knowledge_base/data_structures/kb_link_mount_table.py
class KB_Link_Mount_Table:
    def __init__(self, conn, cursor, database):
        """
        Initialize the LinkMountTable class.
        
        Args:
            conn: Database connection object
            cursor: Database cursor object
            base_table: Base name for the table (will be suffixed with '_link_mount')
        """
        self.conn = conn
        self.cursor = cursor
        self.base_table = database + "_link_mount"
    
    def find_records_by_link_name(self, link_name, kb=None):
        """
        Find records by link_name, optionally filtered by parent_node_kb.
        
        Args:
            link_name (str): The link name to search for
            kb (str, optional): Knowledge base to filter by. If None, search all.
            
        Returns:
            list: List of matching records as dictionaries
        """
        if kb is None:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE link_name = %s
            """
            self.cursor.execute(query, (link_name,))
        else:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE link_name = %s AND knowledge_base = %s
            """
            self.cursor.execute(query, (link_name, kb))
        
        # Get column names from cursor description
        columns = [desc[0] for desc in self.cursor.description]
        # Convert rows to dictionaries
        return [dict(zip(columns, row)) for row in self.cursor.fetchall()]
    
    def find_records_by_mount_path(self, mount_path, kb=None):
        """
        Find records by mount_path (mount_path)), optionally filtered by knowledge_base.
        
        Args:
            mount_path (str): The mount path to search for
            kb (str, optional): Knowledge base to filter by. If None, search all.
            
        Returns:
            list: List of matching records as dictionaries
        """
        
        
        if kb is None:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE mount_path = %s
            """
            self.cursor.execute(query, (mount_path,))
        else:
            query = f"""
                SELECT *
                FROM {self.base_table}
                WHERE mount_path = %s AND knowledge_base = %s
            """
            self.cursor.execute(query, (mount_path, kb))
        
        # Get column names from cursor description
        
        # Convert rows to dictionaries
        return self.cursor.fetchall()
    
    def find_all_link_names(self):
        """
        Get all unique link names from the table.
        
        Returns:
            list: List of all unique link names
        """
        query = f"SELECT DISTINCT link_name FROM {self.base_table} ORDER BY link_name"
        self.cursor.execute(query)
        return [row["link_name"] for row in self.cursor.fetchall()]
    
    def find_all_mount_paths(self):
        """
        Get all unique parent paths (nodes) from the table.
        
        Returns:
            list: List of all unique parent paths
        """
        query = f"SELECT DISTINCT mount_path FROM {self.base_table} ORDER BY mount_path"
        self.cursor.execute(query)
        return [row["mount_path"] for row in self.cursor.fetchall()]

# Example usage:
"""
import psycopg2

# Connect to database
conn = psycopg2.connect(
    host="localhost",
    database="your_db",
    user="your_user",
    password="your_password"
)
cursor = conn.cursor()

# Create instance
link_mount_table = LinkMountTable(conn, cursor, "my_database")

# Use the methods
records = link_mount_table.find_records_by_link_name("example_link")
node_records = link_mount_table.find_records_by_node("root.docs.folder", kb="kb1")
all_links = link_mount_table.find_all_link_names()
all_nodes = link_mount_table.find_all_node()
"""

File: c_programs_and_containers/build_blocks/knowledge_base/postgres/data_structures/luajit/kb_link_mount_table.lua
--[[
  KB_Link_Mount_Table - Query the knowledge_base_link_mount table.
  LuaJIT port using luadbi-postgresql.

  Usage:
    local KB_Link_Mount_Table = require("kb_link_mount_table")
    local lmt = KB_Link_Mount_Table.new(kb_search, "knowledge_base")
    local recs = lmt:find_records_by_link_name("my_link")
]]

local KB_Link_Mount_Table = {}
KB_Link_Mount_Table.__index = KB_Link_Mount_Table

---------------------------------------------------------------------------
-- Helpers
---------------------------------------------------------------------------

local function esc(val)
  if val == nil then return "NULL" end
  local s = tostring(val)
  s = s:gsub("'", "''")
  return "'" .. s .. "'"
end

---------------------------------------------------------------------------
-- Constructor
---------------------------------------------------------------------------

function KB_Link_Mount_Table.new(kb_search, database)
  local self = setmetatable({}, KB_Link_Mount_Table)
  self.kb_search  = kb_search
  self.dbh        = kb_search:get_connection()
  self.base_table = database .. "_link_mount"
  return self
end

---------------------------------------------------------------------------
-- Queries
---------------------------------------------------------------------------

--- Find records by link_name, optionally filtered by knowledge_base.
function KB_Link_Mount_Table:find_records_by_link_name(link_name, kb)
  local sql_str
  if kb then
    sql_str = string.format(
      "SELECT * FROM %s WHERE link_name = %s AND knowledge_base = %s",
      self.base_table, esc(link_name), esc(kb))
  else
    sql_str = string.format(
      "SELECT * FROM %s WHERE link_name = %s",
      self.base_table, esc(link_name))
  end
  return self.kb_search:_raw_query(sql_str)
end

--- Find records by mount_path, optionally filtered by knowledge_base.
function KB_Link_Mount_Table:find_records_by_mount_path(mount_path, kb)
  local sql_str
  if kb then
    sql_str = string.format(
      "SELECT * FROM %s WHERE mount_path = %s AND knowledge_base = %s",
      self.base_table, esc(mount_path), esc(kb))
  else
    sql_str = string.format(
      "SELECT * FROM %s WHERE mount_path = %s",
      self.base_table, esc(mount_path))
  end
  return self.kb_search:_raw_query(sql_str)
end

--- Get all unique link names.
function KB_Link_Mount_Table:find_all_link_names()
  local sql_str = string.format(
    "SELECT DISTINCT link_name FROM %s ORDER BY link_name", self.base_table)
  local rows = self.kb_search:_raw_query(sql_str)
  local rv = {}
  for _, row in ipairs(rows) do rv[#rv + 1] = row.link_name end
  return rv
end

--- Get all unique mount paths.
function KB_Link_Mount_Table:find_all_mount_paths()
  local sql_str = string.format(
    "SELECT DISTINCT mount_path FROM %s ORDER BY mount_path", self.base_table)
  local rows = self.kb_search:_raw_query(sql_str)
  local rv = {}
  for _, row in ipairs(rows) do rv[#rv + 1] = row.mount_path end
  return rv
end

return KB_Link_Mount_Table

